firebase-segmentation/src/main/java/com/google/firebase/segmentation/remote/SegmentationServiceClient.java

```diff
@@ -14,43 +14,148 @@
 
 package com.google.firebase.segmentation.remote;
 
-import com.google.android.gms.tasks.Task;
-import com.google.android.gms.tasks.Tasks;
-import com.squareup.okhttp.OkHttpClient;
-import java.util.concurrent.Executor;
-import java.util.concurrent.Executors;
+import androidx.annotation.NonNull;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.URL;
+import javax.net.ssl.HttpsURLConnection;
+import org.json.JSONException;
+import org.json.JSONObject;
 
 /** Http client that sends request to Firebase Segmentation backend API. To be implemented */
 public class SegmentationServiceClient {
 
-  private final OkHttpClient httpClient;
-  private final Executor httpRequestExecutor;
+  private static final String FIREBASE_SEGMENTATION_API_DOMAIN =
+      "firebasesegmentation.googleapis.com";
+  private static final String UPDATE_REQUEST_RESOURCE_NAME_FORMAT =
+      "projects/%s/installations/%s/customSegmentationData";
+  private static final String CLEAR_REQUEST_RESOURCE_NAME_FORMAT =
+      "projects/%s/installations/%s/customSegmentationData:clear";
+  private static final String FIREBASE_SEGMENTATION_API_VERSION = "v1alpha";
 
   public enum Code {
     OK,
 
-    SERVER_INTERNAL_ERROR,
+    HTTP_CLIENT_ERROR,
 
-    ALREADY_EXISTS,
+    CONFLICT,
 
-    PERMISSION_DENIED
+    NETWORK_ERROR,
+
+    SERVER_ERROR,
+
+    UNAUTHORIZED,
   }
 
-  public SegmentationServiceClient() {
-    httpClient = new OkHttpClient();
-    httpRequestExecutor = Executors.newFixedThreadPool(4);
+  @NonNull
+  public Code updateCustomInstallationId(
+      long projectNumber,
+      @NonNull String apiKey,
+      @NonNull String customInstallationId,
+      @NonNull String firebaseInstanceId,
+      @NonNull String firebaseInstanceIdToken) {
+    String resourceName =
+        String.format(UPDATE_REQUEST_RESOURCE_NAME_FORMAT, projectNumber, firebaseInstanceId);
+    try {
+      URL url =
+          new URL(
+              String.format(
+                  "https://%s/%s/%s?key=%s",
+                  FIREBASE_SEGMENTATION_API_DOMAIN,
+                  FIREBASE_SEGMENTATION_API_VERSION,
+                  resourceName,
+                  apiKey));
+
+      HttpsURLConnection httpsURLConnection = (HttpsURLConnection) url.openConnection();
+      httpsURLConnection.setDoOutput(true);
+      httpsURLConnection.setRequestMethod("PATCH");
+      httpsURLConnection.addRequestProperty(
+          "Authorization", "FIREBASE_INSTALLATIONS_AUTH " + firebaseInstanceIdToken);
+      httpsURLConnection.addRequestProperty("Content-Type", "application/json");
+      OutputStream os = httpsURLConnection.getOutputStream();
+      try {
+        os.write(
```

Comment:
Additionally you should gzip the stream to save bandwidth, here's an example use of httpurlconnection with resource closing and gzip handling:

https://github.com/firebase/firebase-android-sdk/blob/a7761a4e5a5dbf0656baeeb3a97acf71d99332e6/transport/transport-backend-cct/src/main/java/com/google/android/datatransport/cct/CctTransportBackend.java#L220-L257

Reply:
Done.